File: Array & Hashing/3. Longest Substring Without Repeating Characters/solution.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        d = {}
        start = longest = 0

        for i,c in enumerate(s):
            if c in d and d[c] >= start:
                start = d[c]+1
            else:
                longest = max(longest, i-start+1)
            d[c] = i

        return longest

File: Array & Hashing/3. Longest Substring Without Repeating Characters/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();

        int longest = 0;
        int start = 0;
        for (int i = 0; i < s.length(); i++) {
            if(map.containsKey(s.charAt(i)) && map.get(s.charAt(i)) >= start) {
                // set new start (directly after repeated character)
                start = map.get(s.charAt(i))+1;
            } else {
                // get max length
                longest = (i-start+1 > longest) ? i-start+1 : longest;
            }
            
            // cache
            map.put(s.charAt(i), i);
        }

        return longest;
    }
}